packages/core/src/api/serialization/html/externalHTMLExporter.ts

```diff
@@ -0,0 +1,166 @@
+// TODO: IMO this should be part of the formatConversions file since the custom
+//  serializer is used for all HTML & markdown conversions. I think this would
+//  also clean up testing converting to clean HTML.
+import { DOMSerializer, Fragment, Node, Schema } from "prosemirror-model";
+import {
+  blockToNode,
+  nodeToBlock,
+} from "../../nodeConversions/nodeConversions";
+import { BlockNoteEditor } from "../../../BlockNoteEditor";
+import {
+  Block,
+  BlockSchema,
+  PartialBlock,
+} from "../../../extensions/Blocks/api/blockTypes";
+import { unified } from "unified";
+import rehypeParse from "rehype-parse";
+import { simplifyBlocks } from "../../formatConversions/simplifyBlocksRehypePlugin";
+import rehypeStringify from "rehype-stringify";
+import { doc } from "./shared";
+
+// Used to export BlockNote blocks and ProseMirror nodes to HTML for use outside
+// the editor. Blocks are exported using the `toExternalHTML` method in their
+// `blockSpec`, or `toInternalHTML` if `toExternalHTML` is not defined.
+//
+// The HTML created by this serializer is different to what's rendered by the
+// editor to the DOM. This also means that data is likely to be lost when
+// converting back to original blocks. The differences in the output HTML are:
+// 1. It doesn't include the `blockGroup` and `blockContainer` wrappers meaning
+// that nesting is not preserved for non-list-item blocks.
+// 2. `li` items in the output HTML are wrapped in `ul` or `ol` elements.
+// 3. While nesting for list items is preserved, other types of blocks nested
+// inside a list are un-nested and a new list is created after them.
+// 4. The HTML is wrapped in a single `div` element.
+//
+// The serializer has 2 main methods:
+// `exportBlocks`: Exports an array of blocks to HTML.
+// `exportFragment`: Exports a ProseMirror fragment to HTML. This is mostly
+// useful if you want to export a selection which may not start/end at the
+// start/end of a block.
+export interface ExternalHTMLExporter<BSchema extends BlockSchema> {
+  exportBlocks: (blocks: PartialBlock<BSchema>[]) => Promise<string>;
+  exportProseMirrorFragment: (fragment: Fragment) => Promise<string>;
+}
+
+// TODO: This implementation obviously sucks. For now we're just doing the
```

Comment:
`exact same thing` not true, you're also calling `toExternalHTML` instead of internal

Can you also comment what a solution that "does not suck" (i.e.: a better solution) would look like?

Does it really "suck" or is just suboptimal?

Reply:
I sorted out the code duplication issue that it was causing between `toInternalHTML` and `toExternalHTML`. So now the only issue is the fact that it uses the rehype plugin in `serializeProseMirrorFragment` to convert internal HTML to external, causing the function to be async. After looking at alternative solutions I realized that doing the "cleaning" (un-nesting blocks, wrapping list items in list elements, etc), is only really possible after converting to HTML anyway, else it won't comply with the PM schema. So I think now the main issue is just the async - I'll see if I can get around the issue in Firefox, since that's where it's causing issues.

Reply:
Nvm, there's literally a `unified.processSync` method we can use to run the rehype plugin instead of `unified.process`, which fixes the async issue.